ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/cache/LRUCache.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.datastructures.util.cache;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Cache made around LRU-map with fixed size, removing eldest entries (by added) when the space is
+ * over
+ *
+ * @param <K> Keys type
+ * @param <V> Values type
+ */
+public class LRUCache<K, V> implements Cache<K, V> {
```

Comment:
I was going to suggest we consider using the Guava cache libraries rather than writing our own, but their version of `Cache.get` throws a checked exception to allow the fallback function to throw checked exceptions which makes using the cache kind of awful.

Reply:
Heh, looks like we are on the same camp regarding checked exceptions :)
Never used Guava caches, but we may consider them for more advanced use cases. This one is pretty simple and straightforward.

Reply:
Just wondering, why do you guys not like checked exceptions? To be able to handle a wider array of exceptions at a higher level?

Reply:
Checked exceptions can wind up requiring a lot of boilerplate try/catch code that makes it hard to follow control flow well.  If there's an error condition that really does need to be handled (e.g. block failing to import) then my first preference would be return a success or error result but if that's not feasible for some reason then I would use a checked exception. I wouldn't want to use an unchecked exception for something that really needs to be handled (the fact that most of the spec code throws unchecked exceptions when it fails is slightly terrifying).

Things change a bit when working with `CompletableFuture` because unchecked exceptions get picked up as failed results so well and checked exceptions are even more annoying since methods like `thenApply` take a function and don't allow throwing checked exceptions.

So it's one of those highly context dependent rules...

The most annoying thing about checked exceptions though is when a method declares a checked exception that it really shouldn't ever throw.  Which is what happens with guava's cache `get` method.  It declares a checked exception so that it can allow the fallback function to throw a checked exception, but our *actual* fallback function shouldn't ever throw any exception so we wind up being forced to handle a checked exception that shouldn't ever be thrown so there's really nothing useful we can do once we catch it. Lots of boiler plate code and no actual value.

Reply:
That's interesting and great to know. Thank you very much for the detailed explanation @ajsutton

Reply:
@cemozerr 
From my personal experience the major part of checked exception handlings looks like 
```java
catch (SomeCheckedException e) {
  throw new SomeUncheckedException(e);
}
```
In the best case it's 
```java
catch (SomeCheckedException e) {
  throw new SomeUncheckedException("Error while doing that: " + that, e);
}
```

This is because the major part of exceptions are actually unrecoverable at a lower level from one side and from another side it doesn't make too much sense when a higher level method declares a bunch of irrelevant checked exceptions. 
Agree with @ajsutton that in major cases dealing with unchecked exceptions is just a lot of boilerplate code unfortunately. 

But there were always two opposite camps of developers regarding checked exception and of course they are both right at some point